File: rust/examples/linux/power/ina219/complete/src/main.rs
use linux_embedded_hal::I2cdev;
use periph::chips::power::Ina219Full;

fn main() {
    let i2c_bus: u8 = std::env::var("I2C_BUS").ok().and_then(|v| v.parse().ok()).unwrap_or(1);
    let addr: u8 = std::env::var("I2C_ADDR")
        .ok()
        .and_then(|v| u8::from_str_radix(v.trim_start_matches("0x"), 16).ok())
        .unwrap_or(0x40);

    let dev = I2cdev::new(format!("/dev/i2c-{}", i2c_bus)).expect("open i2c bus");
    let mut chip = Ina219Full::new(dev, addr, 0.1, 2.0).expect("init INA219");

    println!("voltage: {:.3} V", chip.voltage().unwrap());          // Read bus voltage, () → f32 V
    println!("shunt_voltage: {:.6} V", chip.shunt_voltage().unwrap()); // Read shunt voltage, () → f32 V
    println!("current: {:.6} A", chip.current().unwrap());         // Read load current, () → f32 A
    println!("power: {:.6} W", chip.power().unwrap());             // Read power, () → f32 W
    println!("conversion_ready: {}", chip.conversion_ready().unwrap()); // Check conversion done, () → bool
    println!("overflow: {}", chip.overflow().unwrap());             // Check math overflow, () → bool

    chip.configure(1, 3, 0x03, 0x03, 7).unwrap();
                                                                  // Configure ADC, (brng 0–1, pga 0–3, badc 0x0F, sadc 0x0F, mode 0–7) → ()

    chip.shutdown().unwrap();           // Put chip into power-down mode, () → ()
    chip.wake().unwrap();              // Restore previous operating mode, () → ()

    chip.reset().unwrap();             // Reset all registers and re-write calibration, () → ()
}

File: cpp/examples/linux/power/INA219/complete/main.cpp
#include <cstdio>
#include <cstdlib>
#include <unistd.h>
#include "I2CConnectionLinux.h"
#include "INA219.h"

int main() {
    const char* bus_env  = getenv("I2C_BUS");
    const char* addr_env = getenv("I2C_ADDR");
    int     bus  = bus_env  ? atoi(bus_env)       : 1;
    uint8_t addr = addr_env ? (uint8_t)strtol(addr_env, nullptr, 0) : 0x40;
    I2CConnectionLinux connection(bus, addr);

    INA219Full ina(connection, 0.1f, 2.0f);                                 // Create INA219 driver, (connection, r_shunt=0.1 Ω, max_current=2.0 A)
                                                                           // calibration register derived from shunt and maximum current

    printf("V=%.3f V I=%.4f A P=%.4f W\n",
           (double)ina.voltage(), (double)ina.current(), (double)ina.power());  // Bus voltage, () → float V ; current, () → float A ; power, () → float W
    printf("shunt=%.6f V\n", (double)ina.shunt_voltage());                // Shunt voltage, () → float V

    ina.configure(INA219Full::BRNG_16V, INA219Full::PGA_1,
                  INA219Full::ADC_12BIT, INA219Full::ADC_AVG_16,
                  INA219Full::MODE_SHUNT_BUS_CONT);                        // Configure ADC, (brng, pga, badc, sadc, mode) → void
                                                                           // 16 V range, ±40 mV shunt range, 16-sample shunt averaging
    printf("ready=%d overflow=%d\n", ina.conversion_ready(), ina.overflow());  // Conversion-ready flag, () → bool ; math overflow flag, () → bool

    ina.configure(INA219Full::BRNG_32V, INA219Full::PGA_8, INA219Full::ADC_12BIT,
                  INA219Full::ADC_12BIT, INA219Full::MODE_SHUNT_BUS_TRIG);
    ina.trigger();                                                         // Start a single-shot conversion, () → void
                                                                           // only meaningful in a triggered mode
    usleep(2000);
    printf("triggered V=%.3f V\n", (double)ina.voltage());

    ina.shutdown();                                                        // Enter power-down, () → void
    ina.wake();                                                            // Restore the previous mode, () → void
    ina.reset();                                                           // Power-on reset, () → void
                                                                           // re-applies the default configuration and calibration
    return 0;
}
